### ISSUE TITLE
Better result compression

### ISSUE THREAD

**Issue by stefanb** [openaddresses/machine#655]
Result zips can be compressed significantly better if simlar rows are nearby.

This was experienced at https://github.com/openaddresses/openaddresses/pull/3168 , where ordering the csv rows by housenubmer,street,....postcode,id reduced the end zip file size in half.

I propose to order the result by all columns between coordinates and hash, so by: housenubmer, street, unit, city, district, region, postcode, id

If possible also consider moving the unit out of the otherwise most simillar parts of the row (to before housenubmer or after zip).

**Comment by migurski** [comment]
Good suggestions! The hash should stay in — we actually don't use it internally at all, but it’s needed by data consumers such as Mapzen’s Pelias.

**Comment by trescube** [comment]
Yeah, Pelias uses the hash for consistent unique address identifiers across builds.

**Comment by stefanb** [comment]
If hash is something that can be calculated from the given data I would say it is a waste to distribute it with each row, because they are redundant and don't compress well as good hashes produce quite seemingly random, hardly compressible output.

Why bother producing hashes in the first place if they are not explicitly needed by the openaddresses processing? Let consumers generate whatever hashes they need. I am sure this can be coordinated with significant current consumers, such as Mapzen's Pelias. My understanding is that they use is as a globally unique id, which breaks anyway when any of the addresse's details or cache fingerprint changes.

Making the end results smaller would make them more accessible for consumers and cheaper to process (processing and transport times, demanding less storage).

**Comment by migurski** [comment]
Thanks Stefan — I think for the time being, I’d like to keep the hashes as they are. They're generally used as an identifier, and I prefer for them to be completely opaque. We generate them in a predictable way, but it might be desirable to change that, shorten them for download size, or make some other change in the future that keeps their function as a durable unique identifier.

I’d be curious to hear from anyone else here whether download size has been a concern.

**Comment by trescube** [comment]
It's convenient having the hash available to identify addresses (like geonames, Who's on First, and OSM) uniquely so that different projects utilizing OA can refer to an address using a common scheme.

**Comment by migurski** [comment]
@stefanb’s comments make me think that it might be smart to improve the hash, for example by making it more compressible while retaining global uniqueness.

**Comment by migurski** [comment]
@trescube, @iandees, and I talked about this issue and we’re going to break it into three new ones to approach in this order:

1. Enumerate hashes with auto-incrementing ID
2. Sort cached rows
3. _I can’t recall the third change_ 

I’m still not sure if the compression benefits alone are sufficient to make these changes, but there seem to be some useful neighboring benefits.

**Comment by iandees** [comment]
> I can’t recall the third change—help?

There was no third change :) You've got 'em all covered here.

**Comment by andrewharvey** [comment]
> I’d be curious to hear from anyone else here whether download size has been a concern.

I do appreciate smaller downloads. I'd be happy with bzip2 or xz, but I totally understand if it's not worth the extra storage/complexity (since I guess you'd still want to offer zip).

**Comment by stefanb** [comment]
It might also be interesting to provide per-country download option, with a constant URL to the latest data.
